modules/redhat_manifest.py

```diff
@@ -0,0 +1,274 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# (c) 2017, Sean O'Keeffe <seanokeeffe797@gmail.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: redhat_manifest
+short_description: Interact with a Red Hat Satellite Subscription Manifest
+description:
+    - Download and modify a Red Hat Satellite Subscription Manifest
+author:
+    - "Sean O'Keeffe (@sean797)"
+options:
+    name:
+        description:
+           - Manifest Name
+    uuid:
+        description:
+           - Manifest uuid
+    username:
+        description:
+            - Username on Foreman server
+        required: true
+    password:
+        description:
+            - Password for user accessing Foreman server
+        required: true
+    pool_id:
+        description:
+            - Subscription pool_id
+    quantity:
+        description:
+            - quantity of pool_id Subscriptions
+        type: int
+    state:
+        description:
+            - State of the Subscription
+        default: present
+        choices:
+            - present
+            - absent
```

Comment:
`state` should be the redhat_manifest state, don't you think? I expect that if I use `state: absent`, the manifest will be removed from access.redhat.com.

Reply:
If it was the manifest state, then you wouldnt be able to remove a sub from a manifest.

Do you think the name should be `redhat_manifest_sub` or something else? Or maybe if state == absent and there are no other subs in it then we delete the manifest?

Reply:
Yeah, you're not going to like this response, but I think the task should represent the state of the manifest. Perhaps it'd be better to add pools/quantities in a dict like so (state in this example is the manifest state, not subscriptions):

```yaml
  - name: "Create {{ manifest_name }} manifest and add 5 rhel subs and ensure 0 rhev subs"
    redhat_manifest:
      name: "{{ manifest_name }}"
      username: "{{ rhsm_username }}"
      password: "{{ rhsm_password }}"
      subscriptions:
        - pool_id: "{{ rhsm_rhel_pool_id }}"
          quantity: 5
        - pool_id: "{{ rhsm_rhev_pool_id }}"
          quantity: 0
      state: present
      validate_certs: "{{ rhsm_validate_certs }}"
```

Sorry I didn't catch this earlier. We can discuss it more and let others give their input, but afaik every `state` param in the modules in the repo refer to the module's object itself (i.e. organization, product, etc.)

Reply:
@mdellweg @ekohl  I'd like to get your opinions here.

Reply:
Without knowing too much details I think I agree with @akofink. state is at the manifest level so it should ensure the state of a manifest.

Reply:
@akofink I did originally think about doing it like that, but I think its better to use `with_items` if you need to add multiple subs, I think having complex dict type options takes away form the simplicity of Ansible.

Reply:
@sean797 You can still implement it in such a way that you can add each subscription one at a time if you prefer; however, users wishing to give their APIs a break will probably want to make a single API call for all their subscriptions.

> complex dict type options

I think Ansible is simple because it uses yaml instead of a DSL. I think users are quite comfortable with more complex yaml (and this is just a list of dicts), as it's intention is to be human-readable. The complexity more comes in the module implementation itself rather than what users will see, which I don't think is an issue.

Reply:
I am really no expert for redhat subscriptions. But maybe that gives me the naïve viewpoint you are aiming for.
I agree, that the `state` parameter should really be used for the entity described in the module name.
However, if i understand the discussion right, and `quantity: 0` is not enough to delete a subscription, another `state: present|absent` on the subscription level could make the addition/deletion of subscriptions from the manifest more explicit.

Reply:
> another state: present|absent on the subscription level could make the addition/deletion of subscriptions from the manifest more explicit

:+1:

Reply:
Would everyone be okay with renaming the current `state` param to `subscription_state` & adding a `manifest_state` param?

Reply:
I'm fine with it. Perhaps name it `pool_state` for consistency, and just `state` for the manifest state to follow what we're doing in the other ansible modules.